File: kotlin/examples/src/main/kotlin/Benchmark.kt
package gr.elaevents.buffela.examples

import gr.elaevents.buffela.deserialization.deserialize
import gr.elaevents.buffela.examples.authToken.AuthTokenPayload
import gr.elaevents.buffela.serialization.serialize
import kotlin.system.measureTimeMillis

private const val warmupReps = 100_000
private const val measurementReps = 10_000_000

private val serialized = payload.serialize()

private fun serialize() {
    payload.serialize()
}

private fun deserialize() {
    AuthTokenPayload.deserialize(serialized)
}

private fun printMeasuredTime(action: String, timeMs: Long) {
    val repTime = (timeMs.toDouble() / measurementReps).toBigDecimal()
    println("$action took ${repTime.toPlainString()} ms on average")
}

fun main() {
    for (i in 0..warmupReps) serialize()

    val serializationTimeMs = measureTimeMillis {
        for (i in 0..measurementReps) serialize()
    }

    printMeasuredTime("Serialization", serializationTimeMs)


    for (i in 0..warmupReps) deserialize()

    val deserializationTimeMs = measureTimeMillis {
        for (i in 0..measurementReps) deserialize()
    }

    printMeasuredTime("Deserialization", deserializationTimeMs)

}

File: js/examples/benchmark.js
const { performance } = require('node:perf_hooks');

const file = process.argv[2]
if (!file) {
    console.error('Please provide a file')
    process.exit(1)
}


const { serialize, deserialize } = require(file)

const warmupReps = 100_000
const measurementReps = 1_000_000


// Serialization

for (let i = 0; i < warmupReps; i++) serialize()

const serializationStart = performance.now()
for (let i = 0; i < measurementReps; i++) serialize()
const serializationTime = performance.now() - serializationStart

console.log('Serialization took', serializationTime / measurementReps, "ms on average")


// Deserialization

const buffer = serialize()

for (let i = 0; i < warmupReps; i++) deserialize(buffer)

const deserializationStart = performance.now()
for (let i = 0; i < measurementReps; i++) deserialize(buffer)
const deserializationTime = performance.now() - deserializationStart

console.log('Deserialization took', deserializationTime / measurementReps, "ms on average")